li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerImpl.java

```diff
@@ -444,15 +446,38 @@ private boolean proceedClosing() {
     if (isClosed()) {
       return false;
     }
-    delegateLock.writeLock().lock();
-    try {
+    synchronized (closeLock) {
       if (isClosed()) {
-        return false;
+        return false; //someone beat us to it
+      }
+      int holds = delegateLock.getReadHoldCount(); //this is for our thread
+      ReentrantReadWriteLock.ReadLock readLock = delegateLock.readLock();
+      ReentrantReadWriteLock.WriteLock writeLock = delegateLock.writeLock();
+      if (holds > 0) { //do we own a read lock ?
```

Comment:
Is this check needed?

Reply:
i dont want to call readLock.unlock(); if i shouldnt, even if it may work due to some implementation detail